File: tinyagent/agent.py
"""Agent class built on top of the agent loop."""

from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator, Callable
from dataclasses import dataclass
from typing import TypeAlias, TypeGuard

from .agent_loop import agent_loop, agent_loop_continue
from .agent_types import (
    ZERO_USAGE,
    AfterToolCallFn,
    AgentContext,
    AgentEndEvent,
    AgentEvent,
    AgentLoopConfig,
    AgentMessage,
    AgentState,
    AgentTool,
    ApiKeyResolver,
    AssistantMessage,
    BeforeToolCallFn,
    ConvertToLlmFn,
    ImageContent,
    Message,
    MessageUpdateEvent,
    Model,
    ShouldStopAfterTurnFn,
    StreamFn,
    TextContent,
    ThinkingBudgets,
    ThinkingContent,
    ThinkingLevel,
    ToolCallContent,
    ToolResultMessage,
    TransformContextFn,
    UserMessage,
    is_agent_end_event,
    is_message_end_event,
    is_message_start_or_update_event,
    is_tool_execution_end_event,
    is_tool_execution_start_event,
    is_turn_end_event,
)
from .caching import add_cache_breakpoints

ConvertToLlmCallback: TypeAlias = ConvertToLlmFn
TransformContextCallback: TypeAlias = TransformContextFn
ApiKeyResolverCallback: TypeAlias = ApiKeyResolver


def _is_nonempty_str(value: object) -> bool:
    return isinstance(value, str) and bool(value.strip())


def _has_meaningful_content(partial: AgentMessage | None) -> bool:
    """Check if partial message has meaningful content worth saving."""

    if not isinstance(partial, AssistantMessage):
        return False

    for item in partial.content:
        if isinstance(item, ThinkingContent) and _is_nonempty_str(item.thinking):
            return True
        if isinstance(item, TextContent) and _is_nonempty_str(item.text):
            return True
        if isinstance(item, ToolCallContent) and _is_nonempty_str(item.name):
            return True
    return False


def extract_text(message: AgentMessage | None) -> str:
    """Extract concatenated text blocks from an agent/LLM message."""

    if not message:
        return ""
    if not isinstance(message, UserMessage | AssistantMessage | ToolResultMessage):
        return ""

    parts: list[str] = []
    for item in message.content:
        if isinstance(item, TextContent) and isinstance(item.text, str):
            parts.append(item.text)
    return "".join(parts)


def _create_error_message(model: Model, error: Exception, was_aborted: bool) -> AgentMessage:
    """Create an error message for the agent."""

    return AssistantMessage(
        content=[TextContent(text="")],
        api=model.api,
        provider=model.provider,
        model=model.id,
        usage=ZERO_USAGE,
        stop_reason="aborted" if was_aborted else "error",
        error_message=str(error),
        timestamp=int(asyncio.get_event_loop().time() * 1000),
    )


def _is_llm_message(message: AgentMessage) -> TypeGuard[Message]:
    role = message.role
    return role in {"user", "assistant", "tool_result"}


async def default_convert_to_llm(messages: list[AgentMessage]) -> list[Message]:
    """Default convert_to_llm: keep only LLM-compatible messages."""

    return [message for message in messages if _is_llm_message(message)]


def _build_transform_context(
    user_transform: TransformContextCallback | None,
    enable_caching: bool,
) -> TransformContextCallback | None:
    """Build the final transform_context callback, optionally composing caching."""
    if not enable_caching:
        return user_transform
    if user_transform is None:
        return add_cache_breakpoints

    # Compose: run caching transform first, then user transform
    async def _composed(
        messages: list[AgentMessage], signal: asyncio.Event | None
    ) -> list[AgentMessage]:
        messages = await add_cache_breakpoints(messages, signal)
        return await user_transform(messages, signal)

    return _composed


@dataclass
class AgentOptions:
    """Options for configuring the Agent."""

    initial_state: AgentState | None = None
    convert_to_llm: ConvertToLlmCallback | None = None
    transform_context: TransformContextCallback | None = None
    steering_mode: str = "one-at-a-time"  # "all" or "one-at-a-time"
    follow_up_mode: str = "one-at-a-time"  # "all" or "one-at-a-time"
    stream_fn: StreamFn | None = None
    session_id: str | None = None
    get_api_key: ApiKeyResolverCallback | None = None
    thinking_budgets: ThinkingBudgets | None = None
    enable_prompt_caching: bool = False
    before_tool_call: BeforeToolCallFn | None = None
    after_tool_call: AfterToolCallFn | None = None
    should_stop_after_turn: ShouldStopAfterTurnFn | None = None


class Agent:
    """Agent class that uses the agent loop directly."""

    def __init__(self, opts: AgentOptions | None = None):
        if opts is None:
            opts = AgentOptions()

        self._state = (
            AgentState.model_validate(opts.initial_state) if opts.initial_state else AgentState()
        )
        self._listeners: set[Callable[[AgentEvent], None]] = set()
        self._abort_event: asyncio.Event | None = None
        self._convert_to_llm = opts.convert_to_llm or default_convert_to_llm
        self._transform_context = _build_transform_context(
            opts.transform_context, opts.enable_prompt_caching
        )
        self._steering_queue: list[AgentMessage] = []
        self._follow_up_queue: list[AgentMessage] = []
        self._steering_mode: str = opts.steering_mode or "one-at-a-time"
        self._follow_up_mode: str = opts.follow_up_mode or "one-at-a-time"
        self.stream_fn: StreamFn | None = opts.stream_fn
        self._session_id: str | None = opts.session_id
        self.get_api_key: ApiKeyResolverCallback | None = opts.get_api_key
        self._running_prompt: asyncio.Future[None] | None = None
        self._partial_message: AgentMessage | None = None
        self._thinking_budgets: ThinkingBudgets | None = opts.thinking_budgets
        self._before_tool_call = opts.before_tool_call
        self._after_tool_call = opts.after_tool_call
        self._should_stop_after_turn = opts.should_stop_after_turn

    @property
    def session_id(self) -> str | None:
        """Get the current session ID used for provider caching."""

        return self._session_id

    @session_id.setter
    def session_id(self, value: str | None) -> None:
        """Set the session ID for provider caching."""

        self._session_id = value

    @property
    def thinking_budgets(self) -> ThinkingBudgets | None:
        return self._thinking_budgets

    @thinking_budgets.setter
    def thinking_budgets(self, value: ThinkingBudgets | None) -> None:
        self._thinking_budgets = value

    @property
    def state(self) -> AgentState:
        return self._state

    def subscribe(self, fn: Callable[[AgentEvent], None]) -> Callable[[], None]:
        """Subscribe to agent events. Returns an unsubscribe function."""

        self._listeners.add(fn)
        return lambda: self._listeners.discard(fn)

    # State mutators
    def set_system_prompt(self, value: str) -> None:
        self._state.system_prompt = value

    def set_model(self, model: Model) -> None:
        self._state.model = model

    def set_thinking_level(self, level: ThinkingLevel) -> None:
        self._state.thinking_level = level

    def set_steering_mode(self, mode: str) -> None:
        self._steering_mode = mode

    def get_steering_mode(self) -> str:
        return self._steering_mode

    def set_follow_up_mode(self, mode: str) -> None:
        self._follow_up_mode = mode

    def get_follow_up_mode(self) -> str:
        return self._follow_up_mode

    def set_tools(self, tools: list[AgentTool]) -> None:
        self._state.tools = tools

    def replace_messages(self, messages: list[AgentMessage]) -> None:
        self._state.messages = messages.copy()

    def append_message(self, message: AgentMessage) -> None:
        self._state.messages = [*self._state.messages, message]

    def steer(self, message: AgentMessage) -> None:
        """Queue a steering message to interrupt the agent mid-run."""

        self._steering_queue.append(message)

    def follow_up(self, message: AgentMessage) -> None:
        """Queue a follow-up message to be processed after the agent finishes."""

        self._follow_up_queue.append(message)

    def clear_steering_queue(self) -> None:
        self._steering_queue = []

    def clear_follow_up_queue(self) -> None:
        self._follow_up_queue = []

    def clear_all_queues(self) -> None:
        self._steering_queue = []
        self._follow_up_queue = []

    def clear_messages(self) -> None:
        self._state.messages = []

    def abort(self) -> None:
        if self._abort_event:
            self._abort_event.set()

    async def wait_for_idle(self) -> None:
        if self._running_prompt:
            await self._running_prompt

    def reset(self) -> None:
        self._state.messages = []
        self._state.is_streaming = False
        self._state.stream_message = None
        self._state.pending_tool_calls = set()
        self._state.error = None
        self._steering_queue = []
        self._follow_up_queue = []

    def _build_input_messages(
        self,
        input_data: str | AgentMessage | list[AgentMessage],
        images: list[ImageContent] | None = None,
    ) -> list[AgentMessage]:
        """Normalize prompt input into a list of AgentMessage objects."""

        if isinstance(input_data, list):
            return input_data
        if isinstance(input_data, str):
            content: list[TextContent | ImageContent] = [TextContent(text=input_data)]
            if images:
                content.extend(images)
            return [
                UserMessage(
                    content=content,
                    timestamp=int(asyncio.get_event_loop().time() * 1000),
                )
            ]
        return [input_data]

    async def prompt(
        self,
        input_data: str | AgentMessage | list[AgentMessage],
        images: list[ImageContent] | None = None,
    ) -> AgentMessage:
        """Send a prompt and return the final assistant message."""

        before = len(self._state.messages)
        async for _ in self._run(input_data, images):
            pass
        return self._last_new_assistant(before)

    async def prompt_text(
        self,
        input_data: str | AgentMessage | list[AgentMessage],
        images: list[ImageContent] | None = None,
    ) -> str:
        return extract_text(await self.prompt(input_data, images=images))

    def stream(
        self,
        input_data: str | AgentMessage | list[AgentMessage],
        images: list[ImageContent] | None = None,
    ) -> AsyncIterator[AgentEvent]:
        """Stream agent events for a prompt."""

        return self._run(input_data, images)

    def stream_text(
        self,
        input_data: str | AgentMessage | list[AgentMessage],
        images: list[ImageContent] | None = None,
    ) -> AsyncIterator[str]:
        """Stream just the assistant text deltas for a prompt."""

        async def _gen() -> AsyncIterator[str]:
            current = ""
            async for event in self.stream(input_data, images=images):
                if event.type != "message_update" or not isinstance(event, MessageUpdateEvent):
                    continue

                msg_obj = event.message
                if not isinstance(msg_obj, AssistantMessage):
                    continue

                ame = event.assistant_message_event
                if ame and ame.type == "text_delta" and ame.delta:
                    delta = str(ame.delta)
                    current += delta
                    yield delta
                    continue

                new_text = extract_text(msg_obj)
                delta = new_text[len(current) :] if new_text.startswith(current) else new_text
                current = new_text
                if delta:
                    yield delta

        return _gen()

    async def continue_(self) -> AgentMessage:
        """Continue from current context (for retry after overflow)."""

        if self._state.is_streaming:
            raise RuntimeError(
                "Agent is already processing. Wait for completion before continuing.",
            )

        messages = self._state.messages
        if len(messages) == 0:
            raise RuntimeError("No messages to continue from")
        if messages[-1].role == "assistant":
            raise RuntimeError("Cannot continue from message role: assistant")

        before = len(messages)
        async for _ in self._run(None, None):
            pass
        return self._last_new_assistant(before)

    async def _run(
        self,
        input_data: str | AgentMessage | list[AgentMessage] | None,
        images: list[ImageContent] | None = None,
    ) -> AsyncIterator[AgentEvent]:
        """Run the agent loop, applying each event to state and yielding it.

        `input_data=None` continues from the existing context instead of adding
        new prompt messages.
        """

        if self._state.is_streaming:
            raise RuntimeError(
                "Agent is already processing a prompt. Use steer() or follow_up() to queue "
                "messages, or wait for completion."
            )

        model = self._state.model
        if not model:
            raise RuntimeError("No model configured")

        messages = None if input_data is None else self._build_input_messages(input_data, images)

        self._setup_run_state()
        context, config = self._build_loop_context_and_config(model)

        try:
            stream_iter = (
                agent_loop(messages, context, config, self._abort_event, self.stream_fn)
                if messages is not None
                else agent_loop_continue(context, config, self._abort_event, self.stream_fn)
            )

            async for event in stream_iter:
                self._apply_event(event)
                self._emit(event)
                yield event

            self._handle_remaining_partial()

        except Exception as err:  # noqa: BLE001
            was_aborted = bool(self._abort_event and self._abort_event.is_set())
            error_msg = _create_error_message(model, err, was_aborted)
            self.append_message(error_msg)
            self._state.error = str(err)
            end_event = AgentEndEvent(messages=[error_msg])
            self._emit(end_event)
            yield end_event

        finally:
            self._cleanup_run_state()

    def _apply_event(self, event: AgentEvent) -> None:
        """Update agent state in response to a loop event."""

        state = self._state
        if is_message_start_or_update_event(event):
            self._partial_message = event.message
            state.stream_message = event.message
        elif is_message_end_event(event):
            self._partial_message = None
            state.stream_message = None
            if event.message is not None:
                self.append_message(event.message)
        elif is_tool_execution_start_event(event):
            state.pending_tool_calls = state.pending_tool_calls | {event.tool_call_id}
        elif is_tool_execution_end_event(event):
            state.pending_tool_calls = state.pending_tool_calls - {event.tool_call_id}
        elif is_turn_end_event(event):
            message = event.message
            if isinstance(message, AssistantMessage) and message.error_message:
                state.error = message.error_message
        elif is_agent_end_event(event):
            state.is_streaming = False
            state.stream_message = None

    def _last_new_assistant(self, before: int) -> AgentMessage:
        """Return the last assistant message appended after index `before`."""

        for msg in reversed(self._state.messages[before:]):
            if msg.role == "assistant":
                return msg
        raise RuntimeError("No assistant message produced")

    def _setup_run_state(self) -> None:
        loop = asyncio.get_event_loop()
        self._running_prompt = loop.create_future()
        self._abort_event = asyncio.Event()
        self._partial_message = None
        self._state.is_streaming = True
        self._state.stream_message = None
        self._state.error = None

    def _build_loop_context_and_config(self, model: Model) -> tuple[AgentContext, AgentLoopConfig]:
        context = AgentContext(
            system_prompt=self._state.system_prompt,
            messages=self._state.messages.copy(),
            tools=self._state.tools,
        )

        config = AgentLoopConfig(
            model=model,
            convert_to_llm=self._convert_to_llm,
            transform_context=self._transform_context,
            get_api_key=self.get_api_key,
            get_steering_messages=self._get_steering_messages,
            get_follow_up_messages=self._get_follow_up_messages,
            before_tool_call=self._before_tool_call,
            after_tool_call=self._after_tool_call,
            should_stop_after_turn=self._should_stop_after_turn,
        )

        return context, config

    def _handle_remaining_partial(self) -> None:
        partial = self._partial_message
        if partial and _has_meaningful_content(partial):
            self.append_message(partial)
        elif partial and self._abort_event and self._abort_event.is_set():
            raise RuntimeError("Request was aborted")

    def _cleanup_run_state(self) -> None:
        self._state.is_streaming = False
        self._state.stream_message = None
        self._state.pending_tool_calls = set()
        self._abort_event = None
        self._partial_message = None
        if self._running_prompt and not self._running_prompt.done():
            self._running_prompt.set_result(None)
        self._running_prompt = None

    @staticmethod
    def _drain_queue(queue: list[AgentMessage], mode: str) -> list[AgentMessage]:
        count = 1 if mode == "one-at-a-time" else len(queue)
        taken = queue[:count]
        del queue[:count]
        return taken

    async def _get_steering_messages(self) -> list[AgentMessage]:
        return self._drain_queue(self._steering_queue, self._steering_mode)

    async def _get_follow_up_messages(self) -> list[AgentMessage]:
        return self._drain_queue(self._follow_up_queue, self._follow_up_mode)

    def _emit(self, event: AgentEvent) -> None:
        for listener in self._listeners:
            listener(event)

File: tinyagent-rs/src/agent.rs
use crate::agent_loop::{agent_loop, agent_loop_continue};
use crate::agent_options::{AgentLoopConfig, AgentOptions, MessageProvider};
use crate::agent_streaming::{
    agent_event_channel, text_channel, AgentEventStream, AgentTextStream,
};
use crate::agent_types::{
    current_timestamp, extract_text, AgentContext, AgentEvent, AgentMessage, AgentState,
    AssistantMessage, Model, QueueMode, StopReason, TextContent, UserContent, UserMessage,
};
use crate::error::{AgentError, Result};
use futures::future::BoxFuture;
use futures::StreamExt;
use std::sync::{Arc, Mutex};
use tokio::sync::{broadcast, Notify};
use tokio_util::sync::CancellationToken;

#[derive(Debug, Clone)]
pub enum PromptInput {
    Text(String),
    Message(AgentMessage),
    Messages(Vec<AgentMessage>),
}

impl From<&str> for PromptInput {
    fn from(value: &str) -> Self {
        Self::Text(value.to_string())
    }
}

impl From<String> for PromptInput {
    fn from(value: String) -> Self {
        Self::Text(value)
    }
}

impl From<AgentMessage> for PromptInput {
    fn from(value: AgentMessage) -> Self {
        Self::Message(value)
    }
}

impl From<Vec<AgentMessage>> for PromptInput {
    fn from(value: Vec<AgentMessage>) -> Self {
        Self::Messages(value)
    }
}

#[derive(Debug)]
struct AgentInner {
    state: AgentState,
    steering_queue: Vec<AgentMessage>,
    follow_up_queue: Vec<AgentMessage>,
    steering_mode: QueueMode,
    follow_up_mode: QueueMode,
    options: AgentOptions,
    abort: Option<CancellationToken>,
    run_notify: Option<Arc<Notify>>,
}

#[derive(Clone)]
pub struct Agent {
    inner: Arc<Mutex<AgentInner>>,
    subscribers: broadcast::Sender<AgentEvent>,
}

impl Default for Agent {
    fn default() -> Self {
        Self::new(AgentOptions::default())
    }
}

impl Agent {
    pub fn new(options: AgentOptions) -> Self {
        let state = options.initial_state();
        let (subscribers, _receiver) = broadcast::channel(256);

        Self {
            inner: Arc::new(Mutex::new(AgentInner {
                state,
                steering_queue: Vec::new(),
                follow_up_queue: Vec::new(),
                steering_mode: options.steering_mode.clone(),
                follow_up_mode: options.follow_up_mode.clone(),
                options,
                abort: None,
                run_notify: None,
            })),
            subscribers,
        }
    }

    pub fn state(&self) -> AgentState {
        self.inner
            .lock()
            .expect("agent lock poisoned")
            .state
            .clone()
    }

    pub fn subscribe(&self) -> broadcast::Receiver<AgentEvent> {
        self.subscribers.subscribe()
    }

    pub fn set_system_prompt(&self, value: impl Into<String>) {
        self.inner
            .lock()
            .expect("agent lock poisoned")
            .state
            .system_prompt = value.into();
    }

    pub fn set_model(&self, model: Model) {
        self.inner.lock().expect("agent lock poisoned").state.model = Some(model);
    }

    pub fn set_tools(&self, tools: Vec<crate::agent_types::AgentTool>) {
        self.inner.lock().expect("agent lock poisoned").state.tools = tools;
    }

    pub fn replace_messages(&self, messages: Vec<AgentMessage>) {
        self.inner
            .lock()
            .expect("agent lock poisoned")
            .state
            .messages = messages;
    }

    pub fn append_message(&self, message: AgentMessage) {
        self.inner
            .lock()
            .expect("agent lock poisoned")
            .state
            .messages
            .push(message);
    }

    pub fn steer(&self, message: AgentMessage) {
        self.inner
            .lock()
            .expect("agent lock poisoned")
            .steering_queue
            .push(message);
    }

    pub fn follow_up(&self, message: AgentMessage) {
        self.inner
            .lock()
            .expect("agent lock poisoned")
            .follow_up_queue
            .push(message);
    }

    pub fn clear_steering_queue(&self) {
        self.inner
            .lock()
            .expect("agent lock poisoned")
            .steering_queue
            .clear();
    }

    pub fn clear_follow_up_queue(&self) {
        self.inner
            .lock()
            .expect("agent lock poisoned")
            .follow_up_queue
            .clear();
    }

    pub fn clear_all_queues(&self) {
        let mut inner = self.inner.lock().expect("agent lock poisoned");
        inner.steering_queue.clear();
        inner.follow_up_queue.clear();
    }

    pub fn clear_messages(&self) {
        self.inner
            .lock()
            .expect("agent lock poisoned")
            .state
            .messages
            .clear();
    }

    pub fn reset(&self) {
        let mut inner = self.inner.lock().expect("agent lock poisoned");
        inner.state.messages.clear();
        inner.state.is_streaming = false;
        inner.state.stream_message = None;
        inner.state.pending_tool_calls.clear();
        inner.state.error = None;
        inner.steering_queue.clear();
        inner.follow_up_queue.clear();
    }

    pub fn abort(&self) {
        if let Some(token) = self
            .inner
            .lock()
            .expect("agent lock poisoned")
            .abort
            .clone()
        {
            token.cancel();
        }
    }

    pub async fn wait_for_idle(&self) {
        let notify = self
            .inner
            .lock()
            .expect("agent lock poisoned")
            .run_notify
            .clone();

        if let Some(notify) = notify {
            notify.notified().await;
        }
    }

    pub async fn prompt(&self, input: impl Into<PromptInput>) -> Result<AssistantMessage> {
        let stream = self.stream(input)?;
        let messages = stream.result().await?;
        find_last_assistant(&messages).ok_or_else(|| {
            AgentError::Internal("no assistant message produced by prompt".to_string())
        })
    }

    pub async fn prompt_text(&self, input: impl Into<PromptInput>) -> Result<String> {
        let message = self.prompt(input).await?;
        Ok(extract_text(&AgentMessage::Assistant(message)))
    }

    pub fn stream(&self, input: impl Into<PromptInput>) -> Result<AgentEventStream> {
        let prompts = normalize_input(input.into());
        self.start_prompt_stream(prompts)
    }

    pub fn stream_text(&self, input: impl Into<PromptInput>) -> Result<AgentTextStream> {
        let mut event_stream = self.stream(input)?;
        let (text_stream, text_tx) = text_channel();

        tokio::spawn(async move {
            while let Some(event) = event_stream.next().await {
                if let AgentEvent::MessageUpdate {
                    assistant_message_event:
                        Some(crate::agent_types::AssistantMessageEvent::TextDelta { delta, .. }),
                    ..
                } = event
                {
                    let _ = text_tx.send(delta);
                }
            }

            let _ = event_stream.result().await;
            drop(text_tx);
        });

        Ok(text_stream)
    }

    pub async fn continue_(&self) -> Result<AssistantMessage> {
        let stream = self.start_continue_stream()?;
        let messages = stream.result().await?;
        find_last_assistant(&messages).ok_or_else(|| {
            AgentError::Internal("no assistant message produced by continue_".to_string())
        })
    }

    fn start_prompt_stream(&self, prompts: Vec<AgentMessage>) -> Result<AgentEventStream> {
        let (context, config, abort, notify, model) = self.prepare_run()?;
        let low_level = agent_loop(prompts, context, config, Some(abort.clone()));
        self.wrap_low_level_stream(low_level, model, abort, notify)
    }

    fn start_continue_stream(&self) -> Result<AgentEventStream> {
        let (context, config, abort, notify, model) = self.prepare_run()?;
        let low_level = agent_loop_continue(context, config, Some(abort.clone()))?;
        self.wrap_low_level_stream(low_level, model, abort, notify)
    }

    fn prepare_run(
        &self,
    ) -> Result<(
        AgentContext,
        AgentLoopConfig,
        CancellationToken,
        Arc<Notify>,
        Model,
    )> {
        let mut inner = self.inner.lock().expect("agent lock poisoned");
        if inner.state.is_streaming {
            return Err(AgentError::AlreadyStreaming);
        }

        let model = inner
            .state
            .model
            .clone()
            .ok_or(AgentError::NoModelConfigured)?;
        let abort = CancellationToken::new();
        let notify = Arc::new(Notify::new());
        inner.abort = Some(abort.clone());
        inner.run_notify = Some(notify.clone());
        inner.state.is_streaming = true;
        inner.state.stream_message = None;
        inner.state.error = None;
        inner.state.pending_tool_calls.clear();

        let context = AgentContext {
            system_prompt: inner.state.system_prompt.clone(),
            messages: inner.state.messages.clone(),
            tools: inner.state.tools.clone(),
        };
        let config = AgentLoopConfig::from_options(
            model.clone(),
            &inner.options,
            Some(self.queue_reader(true)),
            Some(self.queue_reader(false)),
        );

        Ok((context, config, abort, notify, model))
    }

    fn queue_reader(&self, steering: bool) -> MessageProvider {
        let inner = Arc::clone(&self.inner);

        Arc::new(move || {
            let inner = Arc::clone(&inner);
            Box::pin(async move {
                let mut inner = inner.lock().expect("agent lock poisoned");
                let mode = if steering {
                    inner.steering_mode.clone()
                } else {
                    inner.follow_up_mode.clone()
                };
                let queue = if steering {
                    &mut inner.steering_queue
                } else {
                    &mut inner.follow_up_queue
                };

                match mode {
                    QueueMode::OneAtATime => {
                        if queue.is_empty() {
                            Vec::new()
                        } else {
                            vec![queue.remove(0)]
                        }
                    }
                    QueueMode::All => queue.drain(..).collect(),
                }
            }) as BoxFuture<'static, Vec<AgentMessage>>
        })
    }

    fn wrap_low_level_stream(
        &self,
        mut low_level: AgentEventStream,
        model: Model,
        abort: CancellationToken,
        notify: Arc<Notify>,
    ) -> Result<AgentEventStream> {
        let (stream, event_tx, result_tx) = agent_event_channel();
        let inner = Arc::clone(&self.inner);
        let subscribers = self.subscribers.clone();

        tokio::spawn(async move {
            while let Some(event) = low_level.next().await {
                {
                    let mut inner = inner.lock().expect("agent lock poisoned");
                    handle_agent_event(&mut inner.state, &event);
                }
                let _ = subscribers.send(event.clone());
                let _ = event_tx.send(event);
            }

            let result = low_level.result().await;
            let final_result = match result {
                Ok(messages) => Ok(messages),
                Err(error) => {
                    let error_message = create_error_message(&model, &error, abort.is_cancelled());
                    {
                        let mut inner = inner.lock().expect("agent lock poisoned");
                        inner
                            .state
                            .messages
                            .push(AgentMessage::Assistant(error_message.clone()));
                        inner.state.error = Some(error.to_string());
                    }

                    let end_event = AgentEvent::AgentEnd {
                        messages: vec![AgentMessage::Assistant(error_message.clone())],
                    };
                    let _ = subscribers.send(end_event.clone());
                    let _ = event_tx.send(end_event);
                    Ok(vec![AgentMessage::Assistant(error_message)])
                }
            };

            {
                let mut inner = inner.lock().expect("agent lock poisoned");
                inner.state.is_streaming = false;
                inner.state.stream_message = None;
                inner.abort = None;
                inner.run_notify = None;
            }
            notify.notify_waiters();

            let _ = result_tx.send(final_result);
            drop(event_tx);
        });

        Ok(stream)
    }
}

fn normalize_input(input: PromptInput) -> Vec<AgentMessage> {
    match input {
        PromptInput::Text(text) => vec![AgentMessage::User(UserMessage {
            content: vec![UserContent::Text(TextContent::new(text))],
            timestamp: Some(current_timestamp()),
            ..UserMessage::default()
        })],
        PromptInput::Message(message) => vec![message],
        PromptInput::Messages(messages) => messages,
    }
}

fn create_error_message(model: &Model, error: &AgentError, aborted: bool) -> AssistantMessage {
    AssistantMessage {
        content: vec![Some(crate::agent_types::AssistantContent::Text(
            TextContent::new(String::new()),
        ))],
        stop_reason: Some(if aborted {
            StopReason::Aborted
        } else {
            StopReason::Error
        }),
        api: Some(model.api.clone()),
        provider: Some(model.provider.clone()),
        model: Some(model.id.clone()),
        error_message: Some(error.to_string()),
        timestamp: Some(current_timestamp()),
        ..AssistantMessage::default()
    }
}

fn find_last_assistant(messages: &[AgentMessage]) -> Option<AssistantMessage> {
    messages.iter().rev().find_map(|message| match message {
        AgentMessage::Assistant(message) => Some(message.clone()),
        _ => None,
    })
}

fn handle_agent_event(state: &mut AgentState, event: &AgentEvent) {
    match event {
        AgentEvent::MessageStart { message } | AgentEvent::MessageUpdate { message, .. } => {
            state.stream_message = message.clone();
        }
        AgentEvent::MessageEnd { message } => {
            state.stream_message = None;
            if let Some(message) = message {
                state.messages.push(message.clone());
            }
        }
        AgentEvent::ToolExecutionStart { tool_call_id, .. } => {
            state.pending_tool_calls.insert(tool_call_id.clone());
        }
        AgentEvent::ToolExecutionEnd { tool_call_id, .. } => {
            state.pending_tool_calls.remove(tool_call_id);
        }
        AgentEvent::TurnEnd { message, .. } => {
            if let Some(AgentMessage::Assistant(message)) = message {
                if let Some(error) = &message.error_message {
                    state.error = Some(error.clone());
                }
            }
        }
        AgentEvent::AgentEnd { .. } => {
            state.is_streaming = false;
            state.stream_message = None;
        }
        AgentEvent::AgentStart | AgentEvent::TurnStart | AgentEvent::ToolExecutionUpdate { .. } => {
        }
    }
}
